Review followup: PR #21028 — t2891: add binary validator + auto-heal to active setup_opencode_cli (t2888 wrong-file followup)

## Unaddressed review bot suggestions

PR #21028 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/21028

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/tests/test-tool-install-opencode.sh:76`
- `setup-modules/tool-install.sh:1536`
- `setup-modules/tool-install.sh:1567`
- `setup-modules/tool-install.sh:1586`
- `setup-modules/tool-install.sh:1613`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/tests/test-tool-install-opencode.sh:76`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/21028#discussion_r3143468410)

```diff
+}
+extract_functions
+
+source_extracted() {
+	# shellcheck disable=SC2317
+	print_info() { echo "INFO: $*"; return 0; }
+	# shellcheck disable=SC2317
+	print_success() { echo "OK: $*"; return 0; }
+	# shellcheck disable=SC2317
+	print_warning() { echo "WARN: $*"; return 0; }
+	# shellcheck disable=SC2317
+	setup_prompt() { local _var="$1" _prompt="$2" _default="$3"; eval "$_var='$_default'"; return 0; }
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The `setup_prompt` stub uses `eval` for variable assignment. According to the general rules of this repository, `eval` should be avoided for security and safety. Since this is a Bash script, `printf -v` is a safer and more idiomatic alternative for dynamic variable assignment.
> 
> ```suggestion
> setup_prompt() { local _var; local _prompt; local _default; _var="$1"; _prompt="$2"; _default="$3"; printf -v "$_var" '%s' "$_default"; }
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Do not use 'eval' to execute dynamically constructed commands or assignments due to security risks.
> 2. When declaring and assigning local variables, separate the declaration from the assignment for consistency and exit code safety.
> 3. Functions wrapping commands should propagate exit codes rather than masking them with hardcoded returns.
> </details>

#### gemini-code-assist on `setup-modules/tool-install.sh:1536`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/21028#discussion_r3143468413)

```diff
+	[[ -n "$bin" ]] || return 2
+	command -v "$bin" >/dev/null 2>&1 || return 2
+
+	local v
+	v=$("$bin" --version 2>/dev/null || echo "")
+	[[ -n "$v" ]] || return 2
+
+	# Anthropic claude CLI signature — highest-confidence rejection.
+	[[ "$v" == *"(Claude Code)"* ]] && return 1
+
+	# opencode is at 1.x; any 2.x+ is wrong (claude CLI is 2.1.x).
+	[[ "$v" =~ ^[2-9][0-9]*\. ]] && return 1
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The regex `^[2-9][0-9]*\.` correctly identifies versions starting with digits 2 through 9, but it fails to match versions 10 and above (e.g., `10.0.0`) because they start with `1`. To robustly detect all versions greater than or equal to 2, the regex should be updated to handle multi-digit major versions using POSIX character classes.
> 
> ```suggestion
> [[ "$v" =~ ^([2-9]|[1-9][[:digit:]]+)\. ]] && return 1
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Use POSIX character classes like [[:digit:]] instead of non-standard shortcuts for better portability.
> </details>

#### gemini-code-assist on `setup-modules/tool-install.sh:1567`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/21028#discussion_r3143468414)

```diff
+		installer="bun"
+	elif command -v npm >/dev/null 2>&1; then
+		installer="npm"
+	else
+		print_warning "Neither bun nor npm found — cannot heal OpenCode binary"
+		print_info "Install Node.js or Bun first, then re-run 'aidevops update'"
 		return 0
 	fi
 
-	# Need either bun or npm to install
-	local installer=""
+	if run_with_spinner "Reinstalling OpenCode (heal)" npm_global_install "$install_pkg"; then
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> There is an inconsistency between the detected `installer` and the actual command executed. While the function checks for `bun` and sets the `installer` variable accordingly (lines 1557-1560), it hardcodes `npm_global_install` for the actual operation. This is misleading in logs/prompts and will fail if `bun` is the only available package manager. Consider using a more generic installer helper or ensuring the implementation matches the detected tool.

#### gemini-code-assist on `setup-modules/tool-install.sh:1586`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/21028#discussion_r3143468416)

```diff
+	local new_bin
+	new_bin=$(command -v opencode 2>/dev/null || echo "")
+	if [[ -n "$new_bin" ]] && _setup_validate_opencode_binary "$new_bin"; then
+		local new_v
+		new_v=$("$new_bin" --version 2>/dev/null | head -1 || echo "unknown")
+		print_success "OpenCode CLI: $new_bin ($new_v)"
+		mkdir -p "${HOME}/.aidevops" 2>/dev/null || true
+		printf '%s\n' "$new_bin" >"${HOME}/.aidevops/.opencode-bin-resolved" 2>/dev/null || true
+	else
+		local v_after
+		v_after=$("$new_bin" --version 2>/dev/null | head -1 || echo "<missing>")
+		print_warning "Post-heal validation still failing: '$new_bin' returns '$v_after'"
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> If `new_bin` is empty (which can happen if `command -v` fails), attempting to execute `"$new_bin" --version` will result in a shell error. It is safer to verify that `new_bin` is non-empty and executable before attempting to run it.
> 
> ```suggestion
> local v_after; v_after="<missing>"; [[ -x "$new_bin" ]] && v_after=$("$new_bin" --version | head -1 || echo "unknown"); print_warning "Post-heal validation still failing: '${new_bin:-opencode}' returns '$v_after'"
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Add guard clauses to check if a variable is empty before using it, especially when derived from external commands.
> 2. Avoid redundant error suppression like 2>/dev/null if the file's existence or executability has already been verified.
> 3. Separate local variable declaration from assignment for consistency and safety.
> </details>

#### gemini-code-assist on `setup-modules/tool-install.sh:1613`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/21028#discussion_r3143468418)

```diff
+	# Without this, alex-solovyev's runner — where command -v opencode
+	# resolves to @anthropic-ai/claude-code — silently passes through
+	# this function, leaving t2887's runtime canary to throttle the spam
+	# without ever healing the binary.
+	local current_bin
+	current_bin=$(command -v opencode 2>/dev/null || echo "")
+	local validate_rc=0
+	if [[ -n "$current_bin" ]]; then
+		_setup_validate_opencode_binary "$current_bin" || validate_rc=$?
+	else
+		validate_rc=2
+	fi
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> This conditional block is redundant. The `_setup_validate_opencode_binary` function already handles empty or null input by returning `2` (missing/unrunnable). You can simplify this by calling the validator directly and capturing the exit code.
> 
> ```suggestion
> local validate_rc; _setup_validate_opencode_binary "$current_bin"; validate_rc=$?
> ```
> 
> <details>
> <summary>References</summary>
> 
> 1. Avoid redundant pre-checks before calling a helper script if the helper handles the check internally.
> 2. Capture a command's exit code in a variable instead of using $? directly in conditionals.
> 3. Separate local variable declaration from assignment.
> </details>

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/21028#pullrequestreview-4176911296)

> ## Code Review
> 
> This pull request introduces a validation and force-heal mechanism for the OpenCode CLI to address binary name collisions, particularly with the Claude CLI. It includes a new smoke test and updates the installation logic to verify binary versions and persist resolved paths. The review feedback highlights several areas for improvement: replacing unsafe eval usage with printf -v, correcting a regex that fails to account for major versions 10 and above, resolving an inconsistency where the detected installer is ignored in favor of a hardcoded command, and adding safety guards to prevent shell errors when executing potentially empty variables.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.14.11 automated scan.

--- a/.agents/scripts/setup/modules/tool-install.sh
+++ b/.agents/scripts/setup/modules/tool-install.sh
@@ -1642,7 +1642,7 @@ _setup_validate_opencode_binary() {
 	[[ "$v" == *"(Claude Code)"* ]] && return 1
 
 	# opencode is at 1.x; any 2.x+ is wrong (claude CLI is 2.1.x).
-	[[ "$v" =~ ^[2-9][0-9]*\. ]] && return 1
+	[[ "$v" =~ ^([2-9]|[1-9][[:digit:]]+)\. ]] && return 1
 
 	# Sanity: must look like a semver (X.Y.Z).
 	[[ "$v" =~ ^[0-9]+\.[0-9]+\.[0-9]+ ]] || return 1
@@ -1674,7 +1674,9 @@ _setup_opencode_force_heal() {
 	fi
 
 	local install_timeout="${AIDEVOPS_OPENCODE_INSTALL_TIMEOUT:-180}"
-	if run_with_spinner "Reinstalling OpenCode (heal)" _setup_opencode_timeout_cmd "$install_timeout" npm_global_install "$install_pkg"; then
+	# npm_global_install is the shared generic helper; despite its historic
+	# name, it uses bun when available and falls back to npm.
+	if run_with_spinner "Reinstalling OpenCode via $installer (heal)" _setup_opencode_timeout_cmd "$install_timeout" npm_global_install "$install_pkg"; then
 		print_success "OpenCode reinstalled via $installer"
 	else
 		print_warning "Heal install failed via $installer"
@@ -1692,8 +1694,11 @@ _setup_opencode_force_heal() {
 		printf '%s\n' "$new_bin" >"${HOME}/.aidevops/.opencode-bin-resolved" 2>/dev/null || true
 	else
 		local v_after
-		v_after=$(_setup_opencode_first_line "$(_setup_opencode_version_output "$new_bin" 2>/dev/null || printf '<missing>')")
-		print_warning "Post-heal validation still failing: '$new_bin' returns '$v_after'"
+		v_after="<missing>"
+		if [[ -n "$new_bin" ]] && [[ -x "$new_bin" ]]; then
+			v_after=$(_setup_opencode_first_line "$(_setup_opencode_version_output "$new_bin" 2>/dev/null || printf 'unknown')")
+		fi
+		print_warning "Post-heal validation still failing: '${new_bin:-opencode}' returns '$v_after'"
 		print_info "Check PATH: 'which -a opencode' — npm/bun global bin dir must come first"
 	fi
 	return 0
@@ -1716,11 +1721,7 @@ setup_opencode_cli() {
 	local current_bin
 	current_bin=$(command -v opencode 2>/dev/null || echo "")
 	local validate_rc=0
-	if [[ -n "$current_bin" ]]; then
-		_setup_validate_opencode_binary "$current_bin" || validate_rc=$?
-	else
-		validate_rc=2
-	fi
+	_setup_validate_opencode_binary "$current_bin" || validate_rc=$?
 
 	# Already valid → record + early return.
 	if [[ $validate_rc -eq 0 ]]; then

--- a/.agents/scripts/tests/test-tool-install-opencode.sh
+++ b/.agents/scripts/tests/test-tool-install-opencode.sh
@@ -76,7 +76,14 @@ source_extracted() {
 	# shellcheck disable=SC2317
 	print_warning() { echo "WARN: $*"; return 0; }
 	# shellcheck disable=SC2317
-	setup_prompt() { local _var="$1" _prompt="$2" _default="$3"; eval "$_var='$_default'"; return 0; }
+	setup_prompt() {
+		local _var="$1"
+		local _prompt="$2"
+		local _default="$3"
+		: "$_prompt"
+		printf -v "$_var" '%s' "$_default"
+		return $?
+	}
 	# shellcheck disable=SC2317
 	run_with_spinner() { shift; "$@"; return $?; }
 	# shellcheck disable=SC2317
@@ -126,6 +133,21 @@ chmod +x "$SANDBOX/bin/opencode-claude"
 	echo "$rc"
 ) >"$SANDBOX/out2" 2>&1
 assert_eq "claude shim -> rc=1" "1" "$(tail -1 "$SANDBOX/out2")"
+
+# --- Test 2b: validator rejects multi-digit non-opencode majors ------------
+echo "Test 2b: _setup_validate_opencode_binary rejects major >=10"
+cat >"$SANDBOX/bin/opencode-major10" <<'EOF'
+#!/usr/bin/env bash
+[[ "${1:-}" == "--version" ]] && echo "10.0.0"
+EOF
+chmod +x "$SANDBOX/bin/opencode-major10"
+(
+	source_extracted
+	rc=0
+	_setup_validate_opencode_binary "$SANDBOX/bin/opencode-major10" || rc=$?
+	echo "$rc"
+) >"$SANDBOX/out2b" 2>&1
+assert_eq "major 10 shim -> rc=1" "1" "$(tail -1 "$SANDBOX/out2b")"
 
 # --- Test 3: validator on missing path -------------------------------------
 echo "Test 3: _setup_validate_opencode_binary on missing path"

